// unimported/dist/presets/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const meta_1 = require("../meta");
const fs_1 = require("../fs");
const preset = {
    name: 'node',
    isMatch: ({ packageJson }) => Boolean(packageJson),
    getConfig: ({ packageJson, hasPackage }) => {
        const hasFlow = hasPackage('flow-bin');
        const extensions = ['.js', '.jsx', '.ts', '.tsx'];
        const entry = Array.from(new Set((0, fs_1.resolveFilesSync)([
            packageJson.source,
            './src/index',
            './src/main',
            './index',
            './main',
            packageJson.main,
        ], extensions))).filter(meta_1.typedBoolean);
        return {
            entry,
            extensions,
            flow: hasFlow,
            ignorePatterns: [
                '**/node_modules/**',
                `**/*.tests.{js,jsx,ts,tsx}`,
                `**/*.test.{js,jsx,ts,tsx}`,
                `**/*.spec.{js,jsx,ts,tsx}`,
                `**/tests/**`,
                `**/__tests__/**`,
                `**/*.d.ts`,
            ].filter(meta_1.typedBoolean),
            ignoreUnimported: [],
            ignoreUnresolved: [],
            ignoreUnused: [],
        };
    },
};
exports.default = preset;
